hypershift-operator/controllers/nodepool/kubevirt/kubevirt.go

```diff
@@ -246,3 +250,133 @@ func MachineTemplateSpec(image string, nodePool *hyperv1.NodePool) *capikubevirt
 		},
 	}
 }
+
+func CacheImage(ctx context.Context, cl client.Client, hostedClusterNamespace string, bootImage string, imageHash string, nodePool *hyperv1.NodePool) error {
+	kvPlatform := nodePool.Spec.Platform.Kubevirt
+	dvName := getCachedPVCName(bootImage)
+
+	pvc := corev1.PersistentVolumeClaim{}
+
+	err := cl.Get(ctx, client.ObjectKey{Name: dvName, Namespace: hostedClusterNamespace}, &pvc)
+
+	if err == nil { // PVC exists. check if we need to upgrade it
+		return updateCachedImage(ctx, cl, &pvc, bootImage, imageHash, kvPlatform)
+	}
+
+	if !errors.IsNotFound(err) {
+		return fmt.Errorf("failed to get the boot image cache PVC; %w", err)
+	}
+
+	return createDVForCache(ctx, cl, hostedClusterNamespace, bootImage, imageHash, kvPlatform, dvName)
+}
+
+func createDVForCache(ctx context.Context, cl client.Client, hostedClusterNamespace string, bootImage string, imageHash string, kvPlatform *hyperv1.KubevirtNodePoolPlatform, dvName string) error {
+	dv := buildDVForCache(kvPlatform, bootImage, imageHash, dvName, hostedClusterNamespace)
+
+	err := cl.Create(ctx, dv)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create a DataVolume for the boot image cache ; %w", err)
+	}
+
+	return nil
+}
+
+func updateCachedImage(ctx context.Context, cl client.Client, pvc *corev1.PersistentVolumeClaim, bootImage string, imageHash string, kvPlatform *hyperv1.KubevirtNodePoolPlatform) error {
+	if pvc.Annotations[bootImagePVCAnnotationImageHash] == imageHash {
+		return nil // already exists
+	}
+
+	// replace the PVC with a new one
+	err := cl.Delete(ctx, pvc)
+	if err != nil {
+		return fmt.Errorf("failed to delete the boot image cache PVC; %w", err)
+	}
+
+	return createDVForCache(ctx, cl, pvc.Namespace, bootImage, imageHash, kvPlatform, pvc.Name)
```

Comment:
We can't assume that just because a pvc was marked for deletion, that it is actually deleted.

I think it would be wise to give every cached pvc a unique name so we aren't colliding during garbage collection like this.